component/xload/xload.go

```diff
@@ -0,0 +1,300 @@
+	case xcommon.EMode.DOWNLOAD():
+		// Start downloader here
+		err = xl.startDownloader()
+		if err != nil {
+			log.Err("Xload::Start : Failed to start downloader [%s]", err.Error())
+			return err
+		}
+	case xcommon.EMode.UPLOAD():
+		// Start uploader here
+		return fmt.Errorf("uploader is currently unsupported")
+	case xcommon.EMode.SYNC():
+		//Start syncer here
+		return fmt.Errorf("sync is currently unsupported")
+	default:
+		log.Err("Xload::Start : Invalid mode : %s", xl.mode.String())
+		return fmt.Errorf("invalid mode in xload : %s", xl.mode.String())
+	}
+
+	return xl.startComponents()
+}
+
+// Stop : Stop the component functionality and kill all threads started
+func (xl *Xload) Stop() error {
+	log.Trace("Xload::Stop : Stopping component %s", xl.Name())
+
+	xl.comps[0].Stop()
+	xl.blockPool.Terminate()
+
+	// TODO:: xload : should we delete the files from local path
+	err := common.TempCacheCleanup(xl.path)
+	if err != nil {
+		log.Err("unable to clean xload local path [%s]", err.Error())
+	}
+
+	return nil
+}
+
+func (xl *Xload) startDownloader() error {
+	log.Trace("Xload::startDownloader : Starting downloader")
+
+	// Create remote lister pool to list remote files
+	rl, err := comp.NewRemoteLister(xl.path, xl.NextComponent())
+	if err != nil {
+		log.Err("Xload::startDownloader : Unable to create remote lister [%s]", err.Error())
+		return err
+	}
+
+	ds, err := comp.NewDownloadSplitter(xl.blockSize, xl.blockPool, xl.path, xl.NextComponent())
+	if err != nil {
+		log.Err("Xload::startDownloader : Unable to create download splitter [%s]", err.Error())
+		return err
+	}
+
+	rdm, err := comp.NewRemoteDataManager(xl.NextComponent())
+	if err != nil {
+		log.Err("Xload::startUploader : failed to create remote data manager [%s]", err.Error())
+		return err
+	}
+
+	xl.comps = []xcommon.XComponent{rl, ds, rdm}
+	return nil
+}
+
+func (xl *Xload) createChain() error {
+	if len(xl.comps) == 0 {
+		log.Err("Xload::createChain : no component initialized in xload")
+		return fmt.Errorf("no component initialized in xload")
+	}
+
+	currComp := xl.comps[0]
+
+	for i := 1; i < len(xl.comps); i++ {
+		nextComp := xl.comps[i]
+		currComp.SetNext(nextComp)
+		currComp = nextComp
+	}
+
+	return nil
+}
+
+func (xl *Xload) startComponents() error {
+	err := xl.createChain()
+	if err != nil {
+		log.Err("Xload::startComponents : failed to create chain [%s]", err.Error())
+		return err
+	}
+
+	for i := len(xl.comps) - 1; i >= 0; i-- {
+		xl.comps[i].Start()
+	}
+
+	return nil
+}
+
+// ------------------------- Factory -------------------------------------------
+
+// Pipeline will call this method to create your object, initialize your variables here
+func NewXloadComponent() internal.Component {
+	comp := &Xload{}
+	comp.SetName(compName)
+	return comp
+}
+
+// On init register this component to pipeline and supply your constructor
+func init() {
+	internal.AddComponent(compName, NewXloadComponent)
```

Comment:
We shall map the path parameter to tmp-path cli param that file-cache uses so that we dont need to introduce any new cli parameter.

Reply:
Done, reusing the "tmp-path" and "block-size-mb" cli flags for the xload config.